File: classes/listing.js
//Listing Database Schema:
//{_id, user_id, title, description, location, creation_time, expiration_time, price, buy, transaction_id}
//Listings are removed after they expire, or they are removed manually by the user, or they are removed when a transaction based
//on the listing has started, at the point remove all other transactions based on that listing
var Listing = function() {
    function Listing(user_id, title, description, location, expiration_time, price, buy, isbn13) {
        //Listing(id, titile, description, location, creation_time, expiration_time, price, buy)
        //this._id (this will be initialized when listing retrieved from database)
        this.user_id = user_id;
        this.title = title;
        this.description = description;
        this.location = location;
        this.creation_time = new Date().getTime();
        this.expiration_time = expiration_time;
        this.price = price;
        this.buy = buy; //true or false whether this is a buy listing
        // this.transaction_id = null;
        // this.transaction_ids = [];
        this.active = true;

        this.removed_time = null;
        
        this.isbn13 = isbn13

        this.picture_ids = [];
    }

    Listing.prototype = {
        constructor: Listing,
        initFromDatabase: function (listing) {
            this._id = listing._id;
            this.user_id = listing.user_id
            this.title = listing.title;
            this.description = listing.description;
            this.location = listing.location;
            this.creation_time = listing.creation_time;
            this.expiration_time = listing.expiration_time
            this.price = listing.price;
            this.buy = listing.buy;
            // this.transaction_id = listing.transaction_id;
            this.active = listing.active;

            if(listing.first_name != undefined){
                this.first_name = listing.first_name;
            }
            if(listing.last_name != undefined){
                this.last_name = listing.last_name;
            }


            //backward compatability
            if(listing.removed_time != undefined){
                this.removed_time = listing.removed_time;
            }

            if(listing.picture_ids != undefined){
                this.picture_ids = listing.picture_ids;
            }
            if(listing.isbn13 != undefined){
                this.isbn13 = listing.isbn13
            }

        },
        update: function(listing){
            this.title = listing.title;
            this.description = listing.description;
            this.location = listing.location;
            this.expiration_time = listing.expiration_time;
            this.price = listing.price;
            this.buy = listing.buy;
        },
        addTransactionId: function(transaction_id){
            // this.transaction_ids.push(transaction_id.toString());
        },
        removeTransactionId: function(transaction_id){
            // var index = this.transaction_ids.indexOf(transaction_id.toString());
            // if(index != -1){
            //     this.transaction_ids.splice(index, 1)
            // }
        },
        addPictureId: function(picture_id){
            this.picture_ids.push(picture_id.toString());
        },
        removePictureId: function(picture_id){
            var index = this.picture_ids.indexOf(picture_id);
            if(index > -1){
                this.picture_ids.splice(index, 1);
            }
        },
        makeTransaction: function (user_initiate_id) {
            //user_intitate_id is the id of the user that clicks on the listing i.e initiates the transaction
            if (_id == undefined) {
                throw {message: "error, the listing from which we are making a transaction has no id"}
            }
            var user_buy_id = this.buy ? this.user_id : user_initiate_id;
            var user_sell_id = this.buy ? user_initiate_id : this.user_id;
            return new Transaction(user_buy_id, user_sell_id, this._id)
        },
        isExpired: function(){
            if(this.expiration_time <= new Date().getTime()){
                return true;
            }
            return false;
        }

    }

    return Listing;
}();

module.exports = Listing;

File: iOS App/mealplanappiOS/Listing.swift
//
//  Listing.swift
//  mealplanappiOS
//
//  Created by Bowen Jin on 11/11/16.
//  Copyright © 2016 Bowen Jin. All rights reserved.
//

import Foundation

class Listing{
    var _id:String;
    var user_id:String;
    var title:String?;
    var description:String?;
    var location:Location?;
    var creation_time:UInt64;
    var expiration_time:UInt64?;
    var price:Double;
    var buy:Bool?;
    var transaction_id:String?;
    
    var users_name: String?
    
    var picture_ids: [String]?
    
    var first_name: String?
    var last_name: String?
    
    var isbn13: String?
    
    var book: Book?
    
    var buyer_user_ids: [String]?
    
    var active: Bool?
    
    var bought_by_user_id: String?
    
    var sold: Bool?
    
    var time_sold: UInt64?
    
    init(_id:String, user_id:String, title:String, description:String, location: Location, creation_time: UInt64, expiration_time: UInt64, price: Double, buy: Bool, transaction_id:String){
        self._id = _id;
        self.user_id = user_id;
        self.title = title;
        self.description = description;
        self.location = location;
        self.creation_time = creation_time
        self.expiration_time = expiration_time;
        self.price = price
        self.buy = buy;
        self.transaction_id = transaction_id;
    }
    
    init(dictionary: Dictionary<String, Any>){
        self._id = (dictionary["_id"] as? String)!;
        self.user_id = (dictionary["user_id"] as? String)!;
//        self.title = dictionary["title"] as? String;
        self.description = dictionary["description"] as? String;
//        if(dictionary["location"] != nil){
//             self.location = Location(dictionary: (dictionary["location"] as? [String:Any])!);
//        }
        self.creation_time = UInt64(dictionary["creation_time"] as! Double)
//        if(!(dictionary["expiration_time"] is NSNull)){
//            self.expiration_time = UInt64((dictionary["expiration_time"] as? Double)!);
//        }
        self.price = dictionary["price"] as! Double
//        if(!(dictionary["buy"] is NSNull)){
//            self.buy = dictionary["buy"] as? Bool
//        }
        if(dictionary["picture_ids"] != nil && !(dictionary["picture_ids"] is NSNull)){
            self.picture_ids = dictionary["picture_ids"] as! [String]
        }
        
        self.isbn13 = dictionary["isbn13"] as? String
        self.first_name = dictionary["first_name"] as? String
        self.last_name = dictionary["last_name"] as? String
        
        self.book = Book(dictionary: dictionary["book"] as! Dictionary<String, Any>);
        
        self.buyer_user_ids = dictionary["buyer_user_ids"] as? [String]
        
        self.bought_by_user_id = dictionary["bought_by_user_id"] as? String
        
        self.active = dictionary["active"] as? Bool
        
        self.sold = dictionary["sold"] as? Bool
        
        self.time_sold = dictionary["time_sold"] as? UInt64
        //        self.transaction_id = dictionary["transaction_id"] as! String?;
    }
    
    func update(listing: Listing){
        self.title = listing.title
        self.description = listing.description
        self.location = listing.location
        self.expiration_time = listing.expiration_time;
        self.price = listing.price
        self.buy = listing.buy;
        self.picture_ids = listing.picture_ids;
        
    }
}
